Add Implementation Status of WebNN Operations page

ibelem opened the pull request:
Along with a broader introduction to WebNN in upcoming blogs, articles, etc., we need a "Implementation Status of WebNN Operations" page for developers to understand the actual implementation progress. We'll start with the backends or JS ML frameworks we already support. 

1. Add XNNPack/CPU status
2. Add ONNXRuntime WebNN EP
3. Removed formatted markdown for tables based on https://github.com/webmachinelearning/webnn/pull/387

@lisa0314 @miaobin @Honry please help to review the contents of this PR.

@anssiko @huningxin Please provide your feedback if we are OK to share this data or any other thoughs, my plan was to merge this PR when @lisa0314 @miaobin land all the first-wave ops of XNNPack in Chromium in next 1-2 weeks.

![10 239 115 52_4000_webnn-status_ (2)](https://github.com/webmachinelearning/webmachinelearning.github.io/assets/5017359/d4c46307-1e4c-4895-a184-5abc0af130b4)

ibelem committed:
Add implementation status of WebNN Operations page

1. WebNN XNNPack/CPU ops implementation status for first-wave models defined in WebNN spec
2. W3C Web Platform Test (WPT) implementation status
3. TFLite WebNN delegate supported ops status
4. ONNX Runtime Web WebNN EP supported ops status

miaobin commented on webnn-ops.md:
```diff
@@ -0,0 +1,146 @@
+| <span class="spec">[resample2d](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-resample2d)</span> | <span class="x_s">✅</span> static_resize_bilinear_2d | <span class="ep_s">✅</span> Resize |
+| <span class="spec">[reshape](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-reshape)</span> | <span class="x_s">✅</span> static_reshape | <span class="ep_s">✅</span> Reshape |
+| <span class="spec">[sigmoid](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-sigmoid)</span> | <span class="x_s">✅</span> sigmoid | <span class="ep_s">✅</span> Sigmoid |
+| <span class="spec">[split](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-split)</span> | <span class="x_cr">🔍</span> even_split | <span class="ep_wip">🚀</span> Split |
```
WebNN XNNPack backend will support even split by even_split2, even_split3 and even_split4, support uneven split by static_slice.

ibelem replied:
Thanks @miaobin , fixed.

miaobin replied:
LGTM, thanks!

ibelem commented:
@Honry  @huningxin, added TFLite External Delegate status, PTAL.

Honry reviewed (approved):
LGTM, thanks!

anssiko commented:
This is a great starting point, thanks everyone for your contributions!

@ibelem, here are my comments:

- Please investigate if it'd be feasible to use MDN [browser-compat-data](https://github.com/mdn/browser-compat-data) (BCD) conventions for this data to be exported into this presentation style you have here (you'd need to write your own BCD -> md/html exporter). BCD is a machine-readable format that is consumed by MDN and [many other projects](https://github.com/mdn/browser-compat-data/#projects-using-the-data). @dontcallmedom can answer BCD related questions you may have. See also earlier discussion in [#223](https://github.com/webmachinelearning/webnn/issues/223). If this would be too big of an effort for this sprint we could make this BCD alignment a "v2" feature for this implementation status page.

Some minor suggestions:

- Please consider linking to corresponding web-platform-tests for each op (both results https://wpt.fyi/results/webnn and source https://github.com/web-platform-tests/wpt/tree/master/webnn might be useful).

- For permalink, consider /webnn-ops/ -> /webnn-status/ and for the menu link name Operations -> Implementation Status or Impl Status if the longer form does not fit.

This PR fixes [#223](https://github.com/webmachinelearning/webnn/issues/223).

Thanks again for your work on this!

ibelem commented:
- Added web-platform-tests for each op, @brucedai PTAL.
- Updated permalink to [Impl Status] (/webnn-status/)

@anssiko thanks for your great suggestion on the BCD integration, we definitely need to do this for web developers and created an internal task.

The BCD data will cover not only the WebNN ops implementation status but also all the WebNN interfaces, constructors, methods etc., and the browser compatibility data for all of them. Maybe we need to think about what kind of WebNN milestones and what frequency of updates are appropriate.

Go back to this PR:

- The BCD project is a Node.js based project and can be installed via npm, but as you suggested, our webmachinelearning.github.io was built on ruby/jekyll, unable to import the package directly. Do you know if BCD has restful APIs available for other web sites to fetch, rather than install the npm package?
- Per the guidelines of BCD, if we want to have a comprehensive implementation status in one page, we may need to request to fetch several json files and merge them together then convert them into one table. We are not the caniuse.com but the WebML site only, I am afraid the ROI of developing .json files to md or html converter is not good enough. 
  - The WebNN ML interface will be added in api/Navigator.json
  - Many of the json files in api/ folder for WebNN and it's interfaces

WDYT?

anssiko commented:
Thanks for the updates @ibelem!

A good approach might be to:

- merge this PR with its current structure when @lisa0314 @miaobin land all the first-wave ops of XNNPack as suggested. Please ping me when that happens.

- open a new issue for the BCD alignment in this repo and cross-link it with this PR and https://github.com/webmachinelearning/webnn/issues/223 for additional context. Please cc also @dontcallmedom on that BCD issue.

>Do you know if BCD has restful APIs available for other web sites to fetch, rather than install the npm package?

I don't, but @dontcallmedom might know.

ibelem commented:
@lisa0314 @miaobin @Honry please help to review the updated ops support status of this PR.

Honry commented:
@ibelem, for WebNN delegate, we've supported Slice and StridedSlice.

ibelem commented:
@anssiko All the 30 first-wave ops for WebNN XNNPack/CPU backend have been landed in upstream Chromium (yesterday), also updated the TFLite WebNN Delegate and ONNXRuntime WebNN EP supported ops to latest, PTAL.

BruceDai commented:
LGTM, thanks @ibelem

anssiko commented:
Fantastic work @ibelem @lisa0314 @miaobin @Honry and others! :rocket:

I'll announce this to the WG in our upcoming meeting. I expect you'll to keep this important resource up to date tracking both updates to the spec and implementations.

anssiko commented:
@ibelem would it help you to keep the commit history or would you prefer squash and merge?

ibelem committed:
Merge branch 'main' into webnn-ops

ibelem commented:
Thanks @anssiko and @Honry @lisa0314 @miaobin @BruceDai for the review, the commits have been squashed as following:

https://github.com/webmachinelearning/webmachinelearning.github.io/pull/41/commits/675c5ab352df38d469547c3917287c06fa3a3a12

> Add implementation status of WebNN Operations page
> 1. WebNN XNNPack/CPU ops implementation status for first-wave models defined in WebNN spec
> 2. W3C Web Platform Test (WPT) implementation status
> 3. TFLite WebNN delegate supported ops status
> 4. ONNX Runtime Web WebNN EP supported ops status

anssiko commented:
@ibelem now live at https://webmachinelearning.github.io/webnn-status/ :rocket:

The Jekyll version in use does not seem to understand some of the Jekyll syntax used.

For example, `document.querySelector('#spec_total')` returns `null`. A fix would probably be to convert `{: #spec_total }` to `<span id="spec_total">` and similarly for the rest of the elements on this line: https://github.com/webmachinelearning/webmachinelearning.github.io/blob/main/webnn-status.md?plain=1#LL103C3-L103C19

Please take a look if that would fix the issue.
